gnomad_qc/v4/sample_qc/assign_ancestry.py

```diff
@@ -93,41 +149,140 @@ def prep_ht_for_rf(
     :param seed: Random seed, defaults to 24.
     :param test: Whether RF should run on the test QC MT.
     :param only_train_on_hgdp_tgp: Whether to train RF classifier using only the HGDP and 1KG populations. Default is False.
+    :param v4_population_spike: List of populations to spike into training. Must be in v4_population_spike dictionary. Default is None.
+    :param v3_population_spike: List of populations to spike into training. Must be in v3_population_spike dictionary. Default is None.
     :return Table with input for the random forest.
     """
     pop_pca_scores_ht = ancestry_pca_scores(include_unreleasable_samples, test).ht()
+    global joint_qc_meta
+    global v3_meta
+
     joint_meta = joint_qc_meta.ht()[pop_pca_scores_ht.key]
 
-    # TODO: Add code for subpopulations
+    hgdp_tgp_outliers = hgdp_tgp_pop_outliers.ht()
+    hgdp_tgp_outliers = hgdp_tgp_outliers.s.collect()
 
-    # Either train the RF with only HGDP and TGP or all known populations and
-    # previously inferred pops in v3
+    # TODO: Add code for subpopulations
+    # Either train the RF with only HGDP and TGP, or HGDP and TGP and all v2 known labels
     if only_train_on_hgdp_tgp:
         logger.info("Training using HGDP and 1KG samples only...")
         training_pop = hl.or_missing(
-            (joint_meta.v3_meta.v3_subsets.hgdp | joint_meta.v3_meta.v3_subsets.tgp),
+            (joint_meta.v3_meta.v3_subsets.hgdp | joint_meta.v3_meta.v3_subsets.tgp)
+            & (joint_meta.v3_meta.v3_project_pop != "oth")
+            & (~hl.literal(hgdp_tgp_outliers).contains(pop_pca_scores_ht.s)),
             joint_meta.v3_meta.v3_project_pop,
         )
     else:
-        logger.info("Training using all v3 non-oth samples...")
-        training_pop = (
-            hl.case()
-            .when(
-                hl.is_defined(joint_meta.v3_meta.v3_project_pop),
-                joint_meta.v3_meta.v3_project_pop,
-            )
-            .when(
-                joint_meta.v3_meta.v3_population_inference.pop != "oth",
-                joint_meta.v3_meta.v3_population_inference.pop,
-            )
-            .or_missing()  # TODO: Do we want a case for v2 known pops after v3 inferred?
-        )  # TODO: Analysis of where v2_pop does not agree with v3 assigned pop
+        logger.info(
+            "Training using HGDP and 1KG samples and all v2 samples with known pops..."
+        )
+        training_pop = hl.if_else(
+            (joint_meta.v3_meta.v3_subsets.hgdp | joint_meta.v3_meta.v3_subsets.tgp)
+            & (joint_meta.v3_meta.v3_project_pop != "oth")
+            & (~hl.literal(hgdp_tgp_outliers).contains(pop_pca_scores_ht.s)),
+            joint_meta.v3_meta.v3_project_pop,
```

Comment:
This is all the same as above so you can just move the above outside of the if statement, then make this like an `if include_v2_known_in_training:` instead of having a `only_train_on_hgdp_tgp` parameter. Then do what you do below with 
```            .when(
                hl.is_defined(pop_pca_scores_ht.training_pop),
                pop_pca_scores_ht.training_pop,
            )
```

This will need to be moved up too, or down to the bottom and only change the `training_pop` expression not the `pop_pca_scores_ht` . 
```
    pop_pca_scores_ht = pop_pca_scores_ht.annotate(
        training_pop=training_pop,
        hgdp_or_tgp=joint_meta.v3_meta.v3_subsets.hgdp
        | joint_meta.v3_meta.v3_subsets.tgp,
    )
```